main.go

```diff
@@ -2,15 +2,131 @@
 
-type DotnetSdk struct{}
+const (
+	ModSourceDirPath = "/src"
+)
+
+var (
+	IgnorePaths = []string{
+		"**/introspection.json",
+		"**/bin",
+		"**/obj",
+	}
+)
+
+func New(
+	// +optional
+	sdkSourceDir *Directory,
+) *DotnetSdk {
+	if sdkSourceDir == nil {
+		sdkSourceDir = dag.CurrentModule().Source().Directory("sdk")
+	}
+	return &DotnetSdk{
+		SDKSourceDir:  sdkSourceDir,
+		RequiredPaths: []string{},
+		Container:     dag.Container(),
+	}
+}
+
+type DotnetSdk struct {
+	SDKSourceDir  *Directory
+	RequiredPaths []string
+	Container     *Container
+}
+
+func (m *DotnetSdk) ModuleRuntime(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*Container, error) {
+	return m.Container, nil
+}
+
+func (m *DotnetSdk) Codegen(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*GeneratedCode, error) {
+	modName, err := modSource.ModuleName(ctx)
+	if err != nil {
+		return nil, err
+	}
+	subpath, err := modSource.SourceSubpath(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	source := m.
+		WithBase(modSource.ContextDirectory(), subpath).
+		WithSln(modName).
+		WithSdk(subpath, introspectionJson).
+		WithProject(subpath, modName).
+		Container.
+		Directory(ModSourceDirPath)
+
+	return dag.GeneratedCode(source).
+		WithVCSGeneratedPaths([]string{"Dagger.SDK*/**"}).
+		WithVCSIgnoredPaths([]string{"Dagger.SDK*/**", "obj", "bin"}), nil
+}
+
+func (m *DotnetSdk) WithBase(contextDir *Directory, subpath string) *DotnetSdk {
+	m.Container = m.Container.
+		From("mcr.microsoft.com/dotnet/sdk:8.0").
+		WithMountedDirectory(ModSourceDirPath, contextDir).
+		WithWorkdir(path.Join(ModSourceDirPath, subpath))
+	return m
+}
+
+func (m *DotnetSdk) WithSln(modName string) *DotnetSdk {
+	name := strcase.ToCamel(modName)
+	m.Container = m.Container.WithExec([]string{"dotnet", "new", "sln", "--name", name})
+	return m
+}
+
+// Installing sdk into subpath.
+func (m *DotnetSdk) WithSdk(subpath string, introspectionJson string) *DotnetSdk {
+	m.Container = m.Container.
+		WithDirectory(
+			"Dagger.SDK",
+			m.SDKSourceDir.Directory("Dagger.SDK"),
+			ContainerWithDirectoryOpts{Exclude: IgnorePaths},
+		).
+		WithNewFile("Dagger.SDK/introspection.json", ContainerWithNewFileOpts{
+			Contents: introspectionJson,
+		}).
+		WithDirectory(
+			"Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator",
+			m.SDKSourceDir.Directory("Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator"),
+			ContainerWithDirectoryOpts{Exclude: IgnorePaths},
+		).
+		WithExec([]string{"dotnet", "sln", "add", "Dagger.SDK"}).
+		WithExec([]string{"dotnet", "sln", "add", "Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator"})
+
+	return m
+}
+
+func (m *DotnetSdk) WithProject(subpath string, modName string) *DotnetSdk {
+	name := strcase.ToCamel(modName)
+	m.Container = m.Container.
+		WithExec([]string{"dotnet", "new", "console", "--framework", "net8.0", "--output", name, "-n", name}).
```

Comment:
I wonder if there's a way to allow the user, to pick an SDK on their machine. 

dotnet --list-sdks will provide a list of installed SDKS (not all users may be able to use .NET 8, maybe they are only allowed LTS versions of SDK).

Reply:
I don't think there's an easy way to configure the framework version. How would this look in future, would we bump SDK code here when .NET 9 comes out, we make .NET 9?

Reply:
I am just thinking here outloud...

Okay so the dagger code will gen .NET 8, imagine I am developer on work machine, and there is a work policy - only use .NET LTS version. If I am to code against APIs, I will modify the generated code to be .NET 6 - maybe I manually downgrade it. In this case, everything should still work because the .NET source Generator is using .NET Standard APIs which is not unique to .NET 8

The only risk we have here is that the OTHER project, may host .NET8 API code we use or language features only available in .NET 8 - in this case, they may need to modify some bits of the code, but in general, the code should be usable across different .NET versions (except .NET FRamework) - So it will likely be compatible with .NET 6,7,8,9 (6 is soon coming to out of support and .NET 8 is LTS)

https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core So we are safe here.

Reply:
I thihnk we should make the `Dagger.SDK` to use `netstandard2` instead of `net8` to make sure the API compatible with the newer version? We already done that in source generator, the SDK should not be hard. And let the user code be .net 8, 9, etc. depends on our dotnet version defined in the docker image.

Reply:
I think I agree, except for... https://devblogs.microsoft.com/dotnet/the-future-of-net-standard/

I think each year, we can revisit and if Microsoft upgrade Roslyn Source Generators, we can then come back to this and make all `.NET<version>`

ALSO, https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md

I think we need to update Source Generator to use Incremental. I think its minor work, seems to be around performance and not "regenerate" each build time unnecessarily.

Reply:
Agree 👍

Reply:
The Dagger.SDK is now netstandard2.0 on main branch. :)